### DISCUSSION TITLE
Generating the entire key (rather than just the id)

### DISCUSSION THREAD

**Discussion by erquhart** [urql-graphql/urql#2946] [Q&A] [answered]
I have a situation where I have multiple types that are actually the same type. In my case it's because Hasura actions don't support reusing existing types, so an action that returns Type needs to define a new typename for it.

Key generators seemed like an ideal and declarative way to handle this, but I can't define the entire key, just the part after `<type>:`, so I'm instead updating the typename in the object before writing to cache. Definitely not a blocker, obviously, but is there a way to generate the entire key?

**Comment by erquhart** [comment]
Just giving this one a bump, doing a lot of workarounds for this because I can't override typenames coming back from the backend. Just curious if anyone has run into this requirement.

**Comment by kitten** [comment, marked as answer]
I'll start off frankly, Graphcache is in some ways opinionated as well — that's why the updates configuration and many more things are kept in one place, and it's up to you to decide how to split this up and structure it (although, it's encouraged to keep things central and to use `cache.link` where possible)

Basically, amongst the opinion that the cache should reflect the "server state" as closely as possible, we also don't allow overlaps of keys.

So basically, it isn't currently possible to do this, and that's by design. If you have multiple types in the backend that are intended to be the same or overlap then the schema design isn't in order.
On a side note, I generally take a bit of an issue with how Hasura doesn't emphasise how important schema design is, which has lead to similar issues being reported to us in the past 😅 

What I can recommend you is to write an exchange that maps results from the API and translates them to the typename you expect. That _does_ mean that your responses will not match your schema, but it is a more explicit way of overriding the types.

An alternative we've considered in the past is to allow arbitrary updater hooks, e.g. updating arbitrary data in the cache when `Type.x` updates. However, this has always turned out to only be necessary when schema design already put severe limitations on the viability of how a cache can deal with the data it receives, i.e. in cases where queries don't deliver the data a UI could reasonably use 😢 

So, unfortunately, for now, I'll have to advice you to either find a solution via schema design (change the schema to match what you're trying to express), or, write a small exchange that remaps data.

---

We actually have a small utility exchange that will get you halfway there: https://github.com/urql-graphql/urql/blob/main/packages/core/src/exchanges/map.ts
You can place this in front of the exchange that gives you results (typically the `fetchExchange`) and use `onResult` to remap your API keys.

That being said, we obviously don't have any external helpers that'll help you in identifying the right types and fields to remap. But I suspect, you're remapping one `__typename` to another. This means that a remapper would use a similar technique as `maskTypename`: https://github.com/urql-graphql/urql/blob/main/packages/core/src/utils/maskTypename.ts

**Reply by erquhart** [reply]
Ah this is amazing context, thank you! Hasura does indeed introduce some issues when working with types, so I appreciate at least having an escape hatch. Thanks for the pointers here.
